### PR TITLE
Add EntityListProvider with hooks/filters

### PR DESCRIPTION
**Pull request by timbonicus**
Introduces a `useEntityListProvider` hook to request catalog entities from the backend with provided filters.

This is inspired by #2791 and @freben's [experimental branch](https://github.com/backstage/backstage/compare/freben/catalog-filter-context?expand=1). The UI closely follows #2791, where the catalog tabs are instead shown as a component `Type` dropdown, which only shows types that exist in the catalog.

catalog-index

There are a few goals captured here:

1. Enable showing **all** catalog entities, not just components (enabled by, but not implemented in this PR).
2. Provide a generic "catalog entity list" functionality that can be used by CatalogPage, ScaffolderPage, ApiExplorerPage, etc (this only converts CatalogPage, for PR size considerations, but I did convert ScaffolderPage locally as a proof of concept).
3. Allow more backend filtering, as a step-wise performance improvement (does not consider pagination yet).
3. Make it easy to customize CatalogPage by minimizing the JSX footprint.
4. Make it easy to add custom filters.

Adds an `EntityListProvider` to wrap the Catalog home page. This provides a list of resolved entities based on provided filters, which can choose to implement either catalog-backend query filters, or frontend function filters.

I inverted control so the provider/hook doesn't know or care whether a particular filter is frontend or backend. This should make it easy to convert frontend filters to backend ones once more advanced queries are supported. Initial filters can be provided, and the filter context is strongly typed but extensible via a generic on the provider and hook.

The footprint of the CatalogPage with these changes becomes:

```jsx
<CatalogLayout>
  <Content>
    <EntityListProvider initialFilters={initialFilters}>
      <ContentHeader title="Components">
        <CreateComponentButton />
        <SupportButton>All your software catalog entities</SupportButton>
      </ContentHeader>
      <div className={styles.contentWrapper}>
        <div>
          <EntityTypePicker />
          <UserListPicker />
          <EntityTagPicker />
        </div>
        <CatalogTable columns={columns} />
      </div>
    </EntityListProvider>
  </Content>
</CatalogLayout>
```

The goal is to make this so small and composable that customization is trivial. The `div` elements are only for layout and could probably be removed also by making the children aware of grid template areas (indirectly, e.g. `useEntityFilterStyles()`).

Replaces #5549 wherein I royally hosed DCO.

Co-authored-by: Mike Lewis <mtlewis@users.noreply.github.com>

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by dtuite** [comment]
Looks great and I believe this will solve a very real UX need.

**Comment by andrewthauer** [comment]
Thanks for taking up reigns on this @timbonicus 😀. I've had no time to work on a large effort like #2791 recently. I've only quickly skimmed over the PR, but the direction seems good so far. Looking forward to this shipping 🎉.

File: plugins/catalog-react/src/components/EntityTagPicker/EntityTagPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      tags: newTags.length ? new EntityTagFilter(newTags) : undefined,
+    });
+  };
+
+  return (
+    <>
+      <Typography variant="subtitle2" className={classes.title}>
+        Tags
+      </Typography>
+      <List disablePadding dense>
```
Comment (kuangp) [inline plugins/catalog-react/src/components/EntityTagPicker/EntityTagPicker.tsx:74]: Hi, I've noticed through our usage that this list can grow in height infinitely as more tags are defined in the catalog causing the entire page to scroll. Is it possible to set some max height on this list so it has its own dedicated scroll bar on overflow? 

or perhaps a design more like the checkbox select list from the api-docs plugin (which I personally think looks cleaner): 
image

also wanted to note that, the api-docs filter also suffers from the too many tags height issue and work in https://github.com/backstage/backstage/pull/5756 might help with resolving the issue on a single shared base component both plugins can reuse

Comment (freben) [reply plugins/catalog-react/src/components/EntityTagPicker/EntityTagPicker.tsx:74]: checkbox select plus quick filter input box at the top = chef's kiss

Comment (timbonicus) [reply plugins/catalog-react/src/components/EntityTagPicker/EntityTagPicker.tsx:74]: Yep, great call. I'll update this.

Comment (timbonicus) [reply plugins/catalog-react/src/components/EntityTagPicker/EntityTagPicker.tsx]: Updated to use multi-select autocomplete:

multi-select-autocomplete

File: plugins/catalog-react/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  constructor(value: UserListFilterKind) {
+    this.value = value;
+  }
+
+  filterEntity(entity: Entity, env: FilterEnvironment): boolean {
+    switch (this.value) {
+      case 'owned':
+        return env.user !== undefined && isOwnerOf(env.user, entity);
+      case 'starred':
+        return env.isStarredEntity(entity);
```
Comment (Rugvip) [inline plugins/catalog-react/src/types.ts:102]: :shakingfist:, uncomposability of hooks and non-react code strikes again.

Perhaps we could just pass in `user` and `isStarredEntity` to the `UserListFilter` instance instead? Possibly doing away with the env altogether. Feels like the `FilterEnvironment` is a thing that might be in the way when extending things atm.

File: .changeset/pink-countries-pump.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+---
+'@backstage/plugin-catalog': minor
```
Comment (Rugvip) [inline .changeset/pink-countries-pump.md:2]: At stability index 1 with this one, so let's provide a bit more guidance around how to migrate existing custom catalog pages, as much as is sensible anyway

File: plugins/catalog-react/src/hooks/useEntityTypeFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        })
+        .then(response => response.items);
+      return items;
+    }
+    return [];
+  }, [kind, catalogApi]);
+
+  useEffect(() => {
+    // Resolve the unique set of types from returned entities; could be optimized by a new endpoint
+    // in the catalog-backend that does this, rather than loading entities with redundant types.
```
Comment (Rugvip) [inline plugins/catalog-react/src/hooks/useEntityTypeFilter.tsx:66]: 👍

File: plugins/catalog-react/src/types.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+   *
+   * @param entity
+   * @param env
+   */
+  filterEntity?: (entity: Entity, env: FilterEnvironment) => boolean;
+};
+
+export class EntityKindFilter implements EntityFilter {
+  private readonly _value: string;
+  constructor(kind: string) {
```
Comment (Rugvip) [inline plugins/catalog-react/src/types.ts:47]: Just a `constructor(readonly kind: string) {}` and get rid of the prop and getter? Or to you want to be sure it's readonly at runtime too?

**Review by Rugvip — COMMENTED** [review]
I like, especially the approach to the backend/frontend filter split 👍 

Couple of nits and thoughts

File: plugins/catalog-react/src/types.ts (hunk 5 above)
Comment (timbonicus) [reply plugins/catalog-react/src/types.ts:47]: Wow, TIL [parameter-properties](https://www.typescriptlang.org/docs/handbook/2/classes.html#parameter-properties). Much nicer, will update!

File: .changeset/pink-countries-pump.md (hunk 3 above)
Comment (timbonicus) [reply .changeset/pink-countries-pump.md:2]: Still need to update this, will take a look tomorrow.

File: plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  const configApi = useApi(configApiRef);
+  const orgName = configApi.getOptionalString('organization.name') ?? 'Company';
+  const filterGroups = getFilterGroups(orgName);
+
+  const { value: user } = useOwnUser();
+  const { isStarredEntity } = useStarredEntities();
+
+  // Static filters; used for generating counts of potentially unselected kinds
+  const ownedFilter = useMemo(
+    () => new UserListFilter('owned', user, isStarredEntity),
```
Comment (Rugvip) [inline plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx:114]: Would separate `OwnedEntitiesFilter` and `StarredEntitiesFilter` make sense at this point? Seems like the only thing that requires the current `UserListFilter` is the title handling, which we could probably skip or handle differently?

Comment (timbonicus) [reply plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx:114]: Hmm, the main reason not to imo is that the `CatalogPage` with `initiallySelectedFilter` would be more awkward to handle if it was split. And the title, as you say, which I’d planned to replace with selected-filter chips but not yet in this PR.

**Review by Rugvip — APPROVED** [review]
Nice! :shipit: ?

**Pull request merged**